Environment/CTP_environment.py

```diff
@@ -4,86 +4,111 @@
         self.add_expensive_edge = add_expensive_edge
-        
+
         # Generate the graph and get origin and goal
-        # Currently, choosing origin and goal for one agent only. 
-        self.agent_graph,self.origin,self.goal = CTP_generator.generate_graph(num_nodes, key, grid_size = grid_size, prop_stoch=prop_stoch)
+        # Currently, choosing origin and goal for one agent only.
+        self.agent_graph, self.origin, self.goal = CTP_generator.generate_graph(
+            num_nodes, key, grid_size=grid_size, prop_stoch=prop_stoch
+        )
 
         # For multiple agents, origin will be an array of size n_agents and goal will be an array that can be of any size
         self.origin = jnp.array([self.origin])
         self.goal = jnp.array([self.goal])
 
         # An action is just the index of the next node to move to.
-        # An action changes agent_pos to a different node. 
+        # An action changes agent_pos to a different node.
         # Jax.lax.cond doesn't work here because spaces. Discrete is not a valid jax type
-        if(self.add_expensive_edge):
+        if self.add_expensive_edge:
             actions = [num_nodes for _ in range(self.num_agents)]
         else:
-            #If action = self.num_nodes, the agent is saying it's not solvable
-            actions = [num_nodes+1 for _ in range(self.num_agents)]
-        self.action_spaces = spaces.MultiDiscrete(actions) 
-        
-    def reset(self,key:chex.PRNGKey) -> tuple[Observation,EnvState]:
+            # If action = self.num_nodes, the agent is saying it's not solvable
+            actions = [num_nodes + 1 for _ in range(self.num_agents)]
+        self.action_spaces = spaces.MultiDiscrete(actions)
+
+    def reset(self, key: chex.PRNGKey) -> tuple[Observation, EnvState]:
         starting_env_state = EnvState(agents_pos=self.origin)
         # Sample the blocking status
         # self.agent_graph is the graph without the blocking status. self.true_graph contains blocking status
-        self.true_graph = CTP_generator.sample_blocking_prob(key,self.agent_graph)
-        self.solvable = CTP_generator.is_solvable(self.true_graph,self.origin.item(),self.goal.item())
-
-        if(self.solvable is False and self.add_expensive_edge is True):
-            # Add an expensive edge. 
-            largest_edge_weight = jnp.max(self.true_graph.edges['weight'])
-            self.true_graph = CTP_generator.add_expensive_edge(self.true_graph,largest_edge_weight*3,self.origin,self.goal)
+        self.true_graph = CTP_generator.sample_blocking_prob(key, self.agent_graph)
+        self.solvable = CTP_generator.is_solvable(
+            self.true_graph, self.origin.item(), self.goal.item()
+        )
+
+        if self.solvable is False and self.add_expensive_edge is True:
+            # Add an expensive edge.
+            largest_edge_weight = jnp.max(self.true_graph.edges["weight"])
+            self.true_graph = CTP_generator.add_expensive_edge(
```

Comment:
I think the expensive edge should always be there even if the graph is solvable without, because it should be a valid policy for the agent to just take that edge in whichever realisation it exists in